osquery/tables/system/darwin/password_policy.cpp

```diff
@@ -0,0 +1,145 @@
+
+    auto dict = CFArrayGetValueAtIndex((CFArrayRef)content, i);
+    const void* description = nullptr;
+    if (CFDictionaryGetValueIfPresent((CFDictionaryRef)dict,
+                                      CFSTR("policyContentDescription"),
+                                      &description) &&
+        description != nullptr) {
+      r["policy_description"] =
+          getPropertiesFromDictionary((CFDictionaryRef)description, "en");
+    }
+    results.push_back(r);
+  }
+}
+
+QueryData genPasswordPolicy(QueryContext& context) {
+  QueryData results;
+
+  CFErrorRef error = nullptr;
+  // Create a node of type LocalNodes with the default OpenDirectory Session
+  auto node = ODNodeCreateWithNodeType(
+      kCFAllocatorDefault, kODSessionDefault, kODNodeTypeLocalNodes, &error);
+  if (node == nullptr) {
+    VLOG(1) << "password_policy: Error creating an OpenDirectory node";
+    return {};
+  }
+
+  /*
+   * policies is a dictionary with `policyCategoryPasswordContent` as an
+   * optional key. The value  of `policyCategoryPasswordContent` is an array of
+   * policy dictionaries that specify the required content of passwords.
+   * Each element of that array, is a dictionary with following keys:
+   * `policyContent`, `policyContentDescription`, and `policyContentDescription`
+   *
+   * (From Apple's docs)
+   *
+   * Note: `ODNodeCopyAccountPolicies` is the only viable API now, which gets
+   * any policies configured for the node.
+   * Similar API in OpenDirectory has been deprecated.
+   */
+
+  // get policies for the node (i.e. the global policy)
+  auto policies = ODNodeCopyAccountPolicies(node, &error);
+  genRowsFromPolicy(policies, results);
+  if (policies != nullptr) {
+    CFRelease(policies);
+  }
+
+  // populate `uids` with the contraint if present, otherwise populate with all
+  // uids from `users` table
+  std::set<std::string> uids;
+  if (context.constraints.at("uid").exists(EQUALS)) {
+    for (const auto& uid : context.constraints.at("uid").getAll(EQUALS)) {
+      uids.insert(uid);
+    }
+  } else {
+    auto users = SQL::selectAllFrom("users");
+    for (const auto& user : users) {
+      uids.insert(user.at("uid"));
+    }
+  }
+
+  // iterate over uids, and get policy for the user
+  for (const auto& uid : uids) {
+    auto uid_string = CFStringCreateWithCString(
+        kCFAllocatorDefault, uid.c_str(), CFStringGetSystemEncoding());
+    auto query = ODQueryCreateWithNode(kCFAllocatorDefault,
+                                       node,
+                                       (CFTypeRef)kODRecordTypeUsers,
+                                       kODAttributeTypeUniqueID,
+                                       kODMatchEqualTo,
+                                       uid_string,
+                                       nullptr,
+                                       0,
+                                       nullptr);
+    auto records = ODQueryCopyResults(query, false, nullptr);
```

Comment:
Do these need to be released?

Reply:
It's not clear exactly which documentation applies to these opendirectory functions, but it seems like they probably do if these docs apply https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html#//apple_ref/doc/uid/20001148-103029

Reply:
@zwass Ah, thanks! `query` doesn't need to be released, but `records` do, thanks for catching this! Updated